subsys/bluetooth/audio/vcs_client.c

```diff
@@ -0,0 +1,935 @@
+/*  Bluetooth VCS Client - Volume Offset Control Client */
+
+/*
+ * Copyright (c) 2020 Bose Corporation
+ * Copyright (c) 2020-2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+
+#include <sys/check.h>
+
+#include <device.h>
+#include <init.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/vcs.h>
+
+#include "vcs_internal.h"
+#include "aics_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_VCS_CLIENT)
+#define LOG_MODULE_NAME bt_vcs_client
+#include "common/log.h"
+
+struct vcs_instance {
+	struct vcs_state state;
+	uint8_t flags;
+
+	uint16_t start_handle;
+	uint16_t end_handle;
+	uint16_t state_handle;
+	uint16_t control_handle;
+	uint16_t flag_handle;
+	struct bt_gatt_subscribe_params state_sub_params;
+	struct bt_gatt_subscribe_params flag_sub_params;
+
+	bool busy;
+	struct vcs_control_vol cp_val;
+	struct bt_gatt_write_params write_params;
+	struct bt_gatt_read_params read_params;
+	struct bt_gatt_discover_params discover_params;
+	struct bt_uuid_16 uuid;
+
+	uint8_t vocs_inst_cnt;
+	struct bt_vocs *vocs[CONFIG_BT_VCS_CLIENT_MAX_VOCS_INST];
+	uint8_t aics_inst_cnt;
+	struct bt_aics *aics[CONFIG_BT_VCS_CLIENT_MAX_AICS_INST];
+};
+
+/* Callback functions */
+static struct bt_vcs_cb *vcs_client_cb;
+
+static struct vcs_instance vcs_insts[CONFIG_BT_MAX_CONN];
+static int vcs_client_common_vcs_cp(struct bt_conn *conn, uint8_t opcode);
+
+bool bt_vcs_client_valid_vocs_inst(struct bt_conn *conn, struct bt_vocs *vocs)
+{
+	uint8_t conn_index;
+
+	if (!conn) {
+		return false;
+	}
+
+	conn_index = bt_conn_index(conn);
+
+	if (!vocs) {
+		return false;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(vcs_insts[conn_index].vocs); i++) {
+		if (vcs_insts[conn_index].vocs[i] == vocs) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool bt_vcs_client_valid_aics_inst(struct bt_conn *conn, struct bt_aics *aics)
+{
+	uint8_t conn_index;
+
+	if (!conn) {
+		return false;
+	}
+
+	conn_index = bt_conn_index(conn);
+
+	if (!aics) {
+		return false;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(vcs_insts[conn_index].aics); i++) {
+		if (vcs_insts[conn_index].aics[i] == aics) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static uint8_t vcs_client_notify_handler(struct bt_conn *conn,
+					 struct bt_gatt_subscribe_params *params,
+					 const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct vcs_instance *vcs_inst = &vcs_insts[bt_conn_index(conn)];
```

Comment:
I did consider adding an __ASSERT for this, but no:
The connection indexes are static and vcs_insts are static as well, so regardless of the conn (unless conn is NULL), then vcs_inst will always be non-null.

Reply:
What happens if the devices is connected to devices A, B and C, with connection indexes 0, 1 and 2, but only A and C has VCS services?

Reply:
Then 1) this (notify) callback will never be called for the connection between the device and B (since it never subscribed), and even if that happens (i.e. the server notifies with a wrong handle or something similar to that), then the `vcs_inst` pointer is still a valid pointer, but where the callbacks are simply not registered.